androidApp/src/main/java/com/example/tasky/android/agenda/viewmodel/AgendaDetailsViewModel.kt

```diff
@@ -0,0 +1,121 @@
+        when (event) {
+            AgendaDetailsScreenEvent.CloseEditText -> _screenStateFlow.update { it.copy(agendaEditTextType = null) }
+            AgendaDetailsScreenEvent.OnCloseClick -> {}
+            AgendaDetailsScreenEvent.OnDeleteClick -> deleteAgendaItem()
+            is AgendaDetailsScreenEvent.OnDescChange -> updateDesc(event.newDesc)
+            AgendaDetailsScreenEvent.OnDescClick -> _screenStateFlow.update { it.copy(agendaEditTextType = AgendaEditTextType.DESCRIPTION) }
+            AgendaDetailsScreenEvent.OnEditClick -> _screenStateFlow.update { it.copy(isEdit = true) }
+            is AgendaDetailsScreenEvent.OnRemindAtChange -> updateRemindAt(event.newRemindAtTime)
+            AgendaDetailsScreenEvent.OnSaveClick -> saveUpdate()
+            is AgendaDetailsScreenEvent.OnStartDateChange -> updateStartDate(event.newDate)
+            is AgendaDetailsScreenEvent.OnStartTimeChange -> updateStartTime(event.newHour, event.newMinute)
+            is AgendaDetailsScreenEvent.OnTitleChange -> updateTitle(event.newTitle)
+            AgendaDetailsScreenEvent.OnTitleClick -> _screenStateFlow.update { it.copy(agendaEditTextType = AgendaEditTextType.TITLE) }
+        }
+    }
+
+    private fun deleteAgendaItem() {
+        viewModelScope.launch {
+            agendaRepository.deleteAgenda(screenStateFlow.value.agendaItem).onSuccess {
+                _isDeleteSuccess.update { true }
+            }
+        }
+    }
+
+    private fun updateDesc(newDesc: String) {
+        _screenStateFlow.update {
+            it.copy(agendaItem = it.agendaItem.copy(description = newDesc))
+        }
+    }
+
+    private fun updateTitle(newTitle: String) {
+        _screenStateFlow.update {
+            it.copy(agendaItem = it.agendaItem.copy(title = newTitle))
+        }
+    }
+
+    private fun updateStartTime(
+        hour: Int,
+        minute: Int,
+    ) {
+        val agendaItem = screenStateFlow.value.agendaItem
+        val timeMilli = hour.hours.toLong(DurationUnit.MILLISECONDS) + minute.minutes.toLong(DurationUnit.MILLISECONDS)
+        val oneDayMilli = 1.days.toLong(DurationUnit.MILLISECONDS)
+        val newDateTime = agendaItem.getStartTime() / oneDayMilli * oneDayMilli + timeMilli
+        val remindAtDiff = agendaItem.getStartTime() - agendaItem.remindAt
+        _screenStateFlow.update {
+            it.copy(agendaItem = it.agendaItem.copy(startTime = newDateTime, remindAt = newDateTime - remindAtDiff))
+        }
+    }
+
+    private fun updateStartDate(newDate: Long) {
+        val agendaItem = screenStateFlow.value.agendaItem
+        val oneDayMilli = 1.days.toLong(DurationUnit.MILLISECONDS)
+        val newDateTime = newDate + agendaItem.getStartTime() % oneDayMilli
+        val remindAtDiff = agendaItem.getStartTime() - agendaItem.remindAt
+        _screenStateFlow.update {
+            it.copy(agendaItem = it.agendaItem.copy(startTime = newDateTime, remindAt = newDateTime - remindAtDiff))
+        }
+    }
```

Comment:
These two functions seem unnecessarily complicated, with much room for mistakes and hard-to-find bugs. What do you think about modelling the start time as `LocalDateTime` or `ZonedDateTime`? It would be easier to work with than the lower level `Long`. You could then simplify these two functions to a single line of code.

```
is AgendaDetailAction.OnStartDateChanged -> {
  state = state.copy(
    time = LocalDateTime.of(action.newDate, state.time.toLocalTime())
  )
}

is AgendaDetailAction.OnStartTimeChanged -> {
  state = state.copy(
    time = LocalDateTime.of(state.time.toLocalDate(), action.newTime)
  )
}
```

Here, we take the `LocalDateTime` received from the date/time picker dialog and then create a new `LocalDateTime` object using our existing one from the current `state`.

Reply:
However, these values are the one that directly received from the date/time picker dialog. Wouldn't changing them to `LocalDateTime` or `ZonedDateTime` in the dialog add extra complexity to the composable?

Reply:
I find using `LocalDateTime` a lot more convenient to work with. It doesn’t have to be complicated. What do you think of this implementation?

```
// /domain/util package
fun Long.toLocalDate(): LocalDate {
    return Instant
        .ofEpochMilli(this)
        .atZone(ZoneId.systemDefault())
        .toLocalDate()
}

fun Long.toLocalDateTime(): LocalDateTime {
    return Instant
        .ofEpochMilli(this)
        .atZone(ZoneId.systemDefault())
        .toLocalDateTime()
}
```

```
Composable
fun TaskyDatePickerDialog(
    onDateSelected: (LocalDate) -> Unit,
    onDismiss: () -> Unit,
    onError: (UiText.StringResource) -> Unit,
    modifier: Modifier = Modifier
) {
    val state = rememberDatePickerState()
    DatePickerDialog(
        onDismissRequest = onDismiss,
        confirmButton = {
            Button(onClick = {
                val selectedDateMillis = state.selectedDateMillis
                if (selectedDateMillis == null) {
                    onError(UiText.StringResource(R.string.failed_to_select_date_please_try_again))
                    return@Button
                }

                // here the conversion is simple and encapsulated
                onDateSelected(selectedDateMillis.toLocalDate())
                onDismiss()
            }) {
                Text(text = "Confirm")
            }
        },
        dismissButton = {
            Button(onClick = onDismiss) {
                Text(text = "Cancel")
            }
        },
        modifier = modifier,
    ) {
        DatePicker(state = state)
    }
}
```